FirebasePerformance/Sources/FPRNanoPbUtils.m

```diff
@@ -0,0 +1,502 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import "FirebasePerformance/Sources/FPRNanoPbUtils.h"
+
+#ifdef TARGET_HAS_MOBILE_CONNECTIVITY
+#import <CoreTelephony/CTCarrier.h>
+#import <CoreTelephony/CTTelephonyNetworkInfo.h>
+#endif
+#import <SystemConfiguration/SystemConfiguration.h>
+
+#import "FirebasePerformance/Sources/Common/FPRConstants.h"
+#import "FirebasePerformance/Sources/FIRPerformance+Internal.h"
+#import "FirebasePerformance/Sources/FPRDataUtils.h"
+#import "FirebasePerformance/Sources/Public/FIRPerformance.h"
+#import "FirebasePerformance/Sources/Timer/FIRTrace+Internal.h"
+#import "FirebasePerformance/Sources/Timer/FIRTrace+Private.h"
+
+#import "FirebasePerformance/Sources/Gauges/CPU/FPRCPUGaugeData.h"
+#import "FirebasePerformance/Sources/Gauges/Memory/FPRMemoryGaugeData.h"
+
+#define BYTES_TO_KB(x) (x / 1024)
+
+static firebase_perf_v1_NetworkRequestMetric_HttpMethod FPRHTTPMethodForString(
+    NSString *methodString);
+static firebase_perf_v1_NetworkConnectionInfo_NetworkType FPRNetworkConnectionInfoNetworkType(void);
+#ifdef TARGET_HAS_MOBILE_CONNECTIVITY
+static firebase_perf_v1_NetworkConnectionInfo_MobileSubtype FPRCellularNetworkType(void);
+#endif
+NSArray<FPRSessionDetails *> *FPRMakeFirstSessionVerbose(NSArray<FPRSessionDetails *> *sessions);
+
+#pragma mark - Nanopb decode and encode helper methods
+
+pb_bytes_array_t *FPREncodeData(NSData *data) {
+  pb_bytes_array_t *pbBytesArray = calloc(1, PB_BYTES_ARRAY_T_ALLOCSIZE(data.length));
+  if (pbBytesArray != NULL) {
+    [data getBytes:pbBytesArray->bytes length:data.length];
+    pbBytesArray->size = (pb_size_t)data.length;
+  }
+  return pbBytesArray;
+}
+
+pb_bytes_array_t *FPREncodeString(NSString *string) {
+  NSData *stringBytes = [string dataUsingEncoding:NSUTF8StringEncoding];
+  return FPREncodeData(stringBytes);
+}
+
+NSData *FPRDecodeData(pb_bytes_array_t *pbData) {
+  NSData *data = [NSData dataWithBytes:&(pbData->bytes) length:pbData->size];
+  return data;
+}
+
+NSString *FPRDecodeString(pb_bytes_array_t *pbData) {
+  NSData *data = FPRDecodeData(pbData);
+  return [NSString stringWithCString:[data bytes] encoding:NSUTF8StringEncoding];
+}
+
+StringToStringMap *_Nullable FPREncodeStringToStringMap(NSDictionary *_Nullable dict) {
+  StringToStringMap *map = calloc(dict.count, sizeof(StringToStringMap));
+  __block NSUInteger index = 0;
+  [dict enumerateKeysAndObjectsUsingBlock:^(NSString *key, NSString *value, BOOL *stop) {
+    map[index].key = FPREncodeString(key);
+    map[index].value = FPREncodeString(value);
+    index++;
+  }];
+  return map;
+}
+
+NSDictionary<NSString *, NSString *> *FPRDecodeStringToStringMap(StringToStringMap *map,
+                                                                 NSInteger count) {
+  NSMutableDictionary<NSString *, NSString *> *dict = [NSMutableDictionary dictionary];
```

Comment:
General recommendation is to use [[NSMutableDictionary alloc] init].

Reply:
Done.